tests/tx_cancel_test.go

```diff
@@ -16,482 +16,486 @@
+
+	signer := types.MakeSigner(bcdata.bc.Config(), bcdata.bc.CurrentHeader().Number)
+	gasPrice := new(big.Int).SetUint64(bcdata.bc.Config().UnitPrice)
+
+	// 1. Insert a value transfer transaction with nonce 0.
+	{
+		var txs types.Transactions
+
+		amount := new(big.Int).SetUint64(1000)
+		values := map[types.TxValueKeyType]interface{}{
+			types.TxValueKeyNonce:    uint64(0),
+			types.TxValueKeyFrom:     *bcdata.addrs[0],
+			types.TxValueKeyTo:       *bcdata.addrs[0],
+			types.TxValueKeyAmount:   amount,
+			types.TxValueKeyGasLimit: gasLimit,
+			types.TxValueKeyGasPrice: gasPrice,
+		}
+		tx, err := types.NewTransactionWithMap(types.TxTypeValueTransfer, values)
+		assert.Equal(t, nil, err)
+
+		err = tx.SignWithKeys(signer, []*ecdsa.PrivateKey{bcdata.privKeys[0]})
+		assert.Equal(t, nil, err)
+
+		txs = append(txs, tx)
+
+		txpool.AddRemotes(txs)
+
+		pending, queued := txpool.Content()
+		assert.Equal(t, 0, len(queued))
+		assert.Equal(t, 1, len(pending))
+		assert.True(t, tx.Equal(pending[*bcdata.addrs[0]][0]))
+	}
+
+	// 2. Insert a value transfer transaction with nonce 0. This should not be replaced.
+	{
+		var txs types.Transactions
+
+		pending, queued := txpool.Content()
+		oldtx := pending[*bcdata.addrs[0]][0]
+
+		amount := new(big.Int).SetUint64(1000)
+		values := map[types.TxValueKeyType]interface{}{
+			types.TxValueKeyNonce:    uint64(0),
+			types.TxValueKeyFrom:     *bcdata.addrs[0],
+			types.TxValueKeyTo:       *bcdata.addrs[1],
+			types.TxValueKeyAmount:   amount,
+			types.TxValueKeyGasLimit: gasLimit,
+			types.TxValueKeyGasPrice: gasPrice,
+		}
+		tx, err := types.NewTransactionWithMap(types.TxTypeValueTransfer, values)
+		assert.Equal(t, nil, err)
+
+		err = tx.SignWithKeys(signer, []*ecdsa.PrivateKey{bcdata.privKeys[0]})
+		assert.Equal(t, nil, err)
+
+		txs = append(txs, tx)
+
+		txpool.AddRemotes(txs)
+
+		pending, queued = txpool.Content()
+		assert.Equal(t, 0, len(queued))
+		assert.Equal(t, 1, len(pending))
+		assert.True(t, oldtx.Equal(pending[*bcdata.addrs[0]][0]))
+	}
+
+	// 3. Insert a TxCancel transaction with nonce 0. This should replace the tx with the same nonce.
+	{
+		var txs types.Transactions
+
+		values := map[types.TxValueKeyType]interface{}{
+			types.TxValueKeyNonce:    uint64(0),
+			types.TxValueKeyFrom:     *bcdata.addrs[0],
+			types.TxValueKeyGasLimit: gasLimit,
+			types.TxValueKeyGasPrice: gasPrice,
+		}
+		tx, err := types.NewTransactionWithMap(types.TxTypeCancel, values)
+		assert.Equal(t, nil, err)
+
+		err = tx.SignWithKeys(signer, []*ecdsa.PrivateKey{bcdata.privKeys[0]})
+		assert.Equal(t, nil, err)
+
+		txs = append(txs, tx)
+
+		txpool.AddRemotes(txs)
+
+		pending, queued := txpool.Content()
+		assert.Equal(t, 0, len(queued))
+		assert.Equal(t, 1, len(pending))
+		assert.True(t, tx.Equal(pending[*bcdata.addrs[0]][0]))
+	}
+
+	// 4. Insert a TxCancel transaction with nonce 0 and different gas limit. This should replace the tx with the same nonce.
```

Comment:
Can you add a test case with TxTypeCancel that has smaller gas limit and same nonce?

Reply:
sorry please ignore this